src/transformations/defmodel.cpp

```diff
@@ -0,0 +1,445 @@
+#ifdef DEBUG_DEFMODEL
+    void log(const std::string &msg) {
+        pj_log(ctx, PJ_LOG_TRACE, "%s", msg.c_str());
+    }
+#endif
+
+  private:
+    EvaluatorIface(const EvaluatorIface &) = delete;
+    EvaluatorIface &operator=(const EvaluatorIface &) = delete;
+};
+
+struct defmodelData {
+    std::unique_ptr<Evaluator<Grid, GridSet, EvaluatorIface>> evaluator{};
+    EvaluatorIface evaluatorIface;
+
+    explicit defmodelData(PJ_CONTEXT *ctx, PJ *cart)
+        : evaluatorIface(ctx, cart) {}
+
+    defmodelData(const defmodelData &) = delete;
+    defmodelData &operator=(const defmodelData &) = delete;
+};
+
+} // namespace
+
+static PJ *destructor(PJ *P, int errlev) {
+    if (nullptr == P)
+        return nullptr;
+
+    auto Q = static_cast<struct defmodelData *>(P->opaque);
+    delete Q;
+    P->opaque = nullptr;
+
+    return pj_default_destructor(P, errlev);
+}
+
+static PJ_COORD forward_4d(PJ_COORD in, PJ *P) {
+    auto *Q = (struct defmodelData *)P->opaque;
+
+    PJ_COORD out;
+    out.xyzt.t = in.xyzt.t;
+
+    if (!Q->evaluator->forward(Q->evaluatorIface, in.xyzt.x, in.xyzt.y,
+                               in.xyzt.z, in.xyzt.t, out.xyzt.x, out.xyzt.y,
+                               out.xyzt.z)) {
+        return proj_coord_error();
+    }
+
+    return out;
+}
+
+static PJ_COORD reverse_4d(PJ_COORD in, PJ *P) {
+    auto *Q = (struct defmodelData *)P->opaque;
+
+    PJ_COORD out;
+    out.xyzt.t = in.xyzt.t;
+
+    if (!Q->evaluator->inverse(Q->evaluatorIface, in.xyzt.x, in.xyzt.y,
+                               in.xyzt.z, in.xyzt.t, out.xyzt.x, out.xyzt.y,
+                               out.xyzt.z)) {
+        return proj_coord_error();
+    }
+
+    return out;
+}
+
+static void reassign_context(PJ *P, PJ_CONTEXT *ctx) {
+    auto *Q = (struct defmodelData *)P->opaque;
+    if (Q->evaluatorIface.ctx != ctx) {
+        Q->evaluator->clearGridCache();
+        Q->evaluatorIface.ctx = ctx;
+    }
+}
+
+PJ *TRANSFORMATION(defmodel, 1) {
+    // Pass a dummy ellipsoid definition that will be overridden just afterwards
+    auto cart = proj_create(P->ctx, "+proj=cart +a=1");
+    if (cart == nullptr)
+        return destructor(P, ENOMEM);
+
+    /* inherit ellipsoid definition from P to Q->cart */
+    pj_inherit_ellipsoid_def(P, cart);
+
+    auto Q = new defmodelData(P->ctx, cart);
+    P->opaque = (void *)Q;
+    P->destructor = destructor;
+    P->reassign_context = reassign_context;
+
+    const char *model = pj_param(P->ctx, P->params, "smodel").s;
+    if (!model) {
+        proj_log_error(P, "defmodel: +model= should be specified.");
+        return destructor(P, PJD_ERR_NO_ARGS);
+    }
+
+    auto file = NS_PROJ::FileManager::open_resource_file(P->ctx, model);
+    if (nullptr == file) {
+        proj_log_error(P, "defmodel: Cannot open %s", model);
+        return destructor(P, PJD_ERR_INVALID_ARG);
+    }
+    file->seek(0, SEEK_END);
+    unsigned long long size = file->tell();
+    if (size > 10 * 1024 * 1024) {
```

Comment:
magic. why is this size too large?

Reply:
There's no reason for the JSON file to be arbitrarily large, so some threshold has to be established to avoid a potential risk of denial of service. Adding comment to make that explicit